packages/opentelemetry-core/src/context/propagation/HttpTraceContext.ts

```diff
@@ -0,0 +1,110 @@
+/**
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  SpanContext,
+  HttpTextFormat,
+  TraceOptions,
+} from '@opentelemetry/types';
+import { TraceState } from '../../trace/TraceState';
+
+export const TRACE_PARENT_HEADER = 'traceparent';
+export const TRACE_STATE_HEADER = 'tracestate';
+const VALID_TRACE_PARENT_REGEX = /^[\da-f]{2}-[\da-f]{32}-[\da-f]{16}-[\da-f]{2}$/;
+const VALID_TRACEID_REGEX = /^[0-9a-f]{32}$/i;
+const VALID_SPANID_REGEX = /^[0-9a-f]{16}$/i;
+const INVALID_ID_REGEX = /^0+$/i;
+const VERSION = '00';
+
+function parse(traceParent: string): SpanContext | null {
+  const match = traceParent.match(VALID_TRACE_PARENT_REGEX);
+  if (!match) return null;
+  const parts = traceParent.split('-');
+  const [version, traceId, spanId, option] = parts;
+  // tslint:disable-next-line:ban Needed to parse hexadecimal.
+  const traceOptions = parseInt(option, 16);
+
+  if (
```

Comment:
This condition is redundant as it is impossible for these to be invalid if the `traceparent` is valid.

Reply:
This is important to check **at least one non-zero byte** in `SpanId` and `TraceId` and currently supported version is `00`.

Reply:
This should be done in the traceparent regex since it means that the entire parent is invalid if any of these fail.

Reply:
Do we have any use-case like this? i.e. valid `traceId` but invalid `spanId` or vice-versa. OR invalid version but valid `traceId` and `spanId`. Let me know if I am missing something.

OR are you hinting about invalid `traceParentHeader` but valid `traceStateHeader`?

Reply:
What I mean is that a single regex can validate both the format and the values. Then the traceparent is either valid or invalid as a whole. Validating a subset of the traceparent after the traceparent has already been validated is not necessary.

Reply:
One of the challenges mentioned in another comment thread here is that a single traceparent regex ends up being pretty massive.

What if once it found an invalid item, it reset the state of the elements it had processed so far?

Reply:
I still think this could be simplified (which would also be more performant) but it can be optimized later.